openshift/machine-config-operator
# MCO-679: MCO-830: Managed boot images MVP

This implements https://github.com/openshift/enhancements/pull/1496, for the GCP platform.

What does it do?
--------------------------------

Adds a new subcontroller `machine_set_boot_image_controller`  to the MCC, that
- Listens on machineset changes or `coreos-bootimages` configmap changes
- Determines architecture and infra of cluster to index the correct boot image
- Updates machineset(s) with newer bootimages if there is a delta after 1 master node has completed the upgrade
- Patches machine set via machineclient

How do I test this?
----------------------------
- Launch a cluster on 4.15. Check the first few lines of the machine config daemon logs, and make a note of the node's [aleph version](https://github.com/coreos/fedora-coreos-tracker/blob/main/internals/README-internals.md#aleph-version). This is the original boot image.
```
I1218 20:35:06.196301    2320 rpm-ostree.go:308] Running captured: rpm-ostree status
I1218 20:35:06.240687    2320 daemon.go:1598] State: idle
Deployments:
* ostree-unverified-registry:registry.ci.openshift.org/ocp/4.16-2023-12-18-045249@sha256:a2900a99595cbf0cf3862d0066d37b2e95a95e032659469d59f616eb9752d3c0
                   Digest: sha256:a2900a99595cbf0cf3862d0066d37b2e95a95e032659469d59f616eb9752d3c0
                  Version: 416.92.202312150201-0 (2023-12-18T16:15:22Z)
I1218 20:35:06.241321    2320 coreos.go:53] CoreOS aleph version: mtime=2023-11-24 16:50:34.214 +0000 UTC
{
   "build": "415.92.202311241643-0",
   "imgid": "rhcos-415.92.202311241643-0-qemu.x86_64.qcow2",
   "ostree-commit": "3aff20eacec06af854303111319e74d9dc84c241af5c57dc8ae3330a8ae5b086",
   "ref": ""
}

```
------------------------
- Upgrade to this PR. This _should_ create a version skew between boot images and current node image. During this upgrade, `coreos-bootimages` configmap will get updated by a CVO manifest to the newest boot images.  
- Turn on the tech preview feature gate by applying this YAML:
```
apiVersion: config.openshift.io/v1
kind: FeatureGate
metadata:
  name: cluster
spec:
  featureSet: TechPreviewNoUpgrade
```
- This should trigger a machineset reconciliation loop. Observe logs on the MCC with this command:
```
oc logs -n openshift-machine-config-operator -f "$(oc get pod -o name -l='k8s-app=machine-config-controller' -n openshift-machine-config-operator)" | grep "machine_set"
```
You should see all the machinesets being patched:
```
I1218 20:31:52.657798       1 machine_set_boot_image_controller.go:143] "FeatureGates changed" enabled=["AdminNetworkPolicy","AlibabaPlatform","AutomatedEtcdBackup","AzureWorkloadIdentity","BuildCSIVolumes","CSIDriverSharedResource","CloudDualStackNodeIPs","DNSNameResolver","DynamicResourceAllocation","ExternalCloudProvider","ExternalCloudProviderAzure","ExternalCloudProviderExternal","ExternalCloudProviderGCP","GCPClusterHostedDNS","GCPLabelsTags","GatewayAPI","InsightsConfigAPI","InstallAlternateInfrastructureAWS","KMSv1","MachineAPIProviderOpenStack","MachineConfigNodes","ManagedBootImages","MaxUnavailableStatefulSet","MetricsServer","MixedCPUsAllocation","NetworkLiveMigration","NodeSwap","OnClusterBuild","OpenShiftPodSecurityAdmission","PrivateHostedZoneAWS","RouteExternalCertificate","SignatureStores","SigstoreImageVerification","VSphereControlPlaneMachineSet","VSphereStaticIPs","ValidatingAdmissionPolicy"] disabled=["ClusterAPIInstall","DisableKubeletCloudCredentialProviders","EventedPLEG","MachineAPIOperatorDisableMachineHealthCheckController"]
I1218 20:31:52.657849       1 machine_set_boot_image_controller.go:156] Trigger a sync as this feature was turned on
I1218 20:31:52.657964       1 machine_set_boot_image_controller.go:560] Reconciling machineset djoshy10-rvsmr-worker-a on GCP, with arch x86_64
I1218 20:31:52.658487       1 machine_set_boot_image_controller.go:560] Reconciling machineset djoshy10-rvsmr-worker-b on GCP, with arch x86_64
I1218 20:31:52.709865       1 machine_set_boot_image_controller.go:585] New target boot image: projects/rhcos-cloud/global/images/rhcos-415-92-202311241643-0-gcp-x86-64
I1218 20:31:52.709888       1 machine_set_boot_image_controller.go:586] Current image: projects/rhcos-cloud/global/images/rhcos-415-92-202311241643-gcp-x86-64
I1218 20:31:52.709961       1 machine_set_boot_image_controller.go:402] Patching machineset djoshy10-rvsmr-worker-a
I1218 20:31:52.711798       1 machine_set_boot_image_controller.go:405] No patching required for machineset djoshy10-rvsmr-worker-b

```
If this doesn't happen, or if logs say that no patching was required for the machinesets, then no skew was found between the boot images and the last upgrade.

- Scale up a new node on any of the machinesets
```
oc scale --replicas=2 machineset <machineset> -n openshift-machine-api
```
- Once the node is up and in the worker pool, observe the MCD logs of the newly scaled up node. At the daemon startup, the [aleph version](https://github.com/coreos/fedora-coreos-tracker/blob/main/internals/README-internals.md#aleph-version) should now indicate the new boot image.
------------------------
Other notes:
Once the feature gate is removed, this reconciliation would take place when at least 1 master node has completed an upgrade(and no other upgrades are pending). This is by design, as one would expect boot image updates to take place during an upgrade.
